File: apps/cli/thecode/sync.py
"""Client de synchronisation.

Le carnet est chiffré **avant** de quitter l'appareil, avec la clef de
transfert dérivée de la clef maîtresse. Le serveur ne reçoit que des blocs
opaques : il ne peut ni lire les sites, ni les identifiants, ni rien déduire
au-delà du nombre d'entrées.

Les identifiants du compte de synchronisation sont volontairement distincts de
la clef maîtresse. S'authentifier avec celle-ci ferait qu'une faiblesse du
service exposerait les mots de passe eux-mêmes.
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .transfer import derive_transfer_key
from .vault import Conflict, drop_version, merge, select_for_push

DEFAULT_ENDPOINT = "https://thecode-api.julsql.fr"


class SyncError(Exception):
    """Échec de synchronisation : réseau, authentification, ou conflit."""


PLAN_FREE = "free"
PLAN_PRO = "pro"


def is_paid_plan(plan: str | None) -> bool:
    """Vrai quand l'offre donne droit au compteur."""
    return plan == PLAN_PRO


@dataclass
class Credentials:
    """Jetons de session. Stockés à part du carnet, et jamais dans le carnet."""

    endpoint: str
    access_token: str
    refresh_token: str
    #: Offre du compte, telle que le service l'a dite la dernière fois.
    #:
    #: Gardée avec les jetons parce que la génération se fait hors ligne : sans
    #: cette trace, la commande ne saurait pas quoi autoriser tant que le
    #: service n'a pas répondu, et autoriserait donc tout.
    plan: str = PLAN_FREE

    @staticmethod
    def path() -> Path:
        base = os.environ.get("XDG_CONFIG_HOME")
        root = Path(base) if base else Path.home() / ".config"
        return root / "thecode" / "session.json"

    def save(self) -> None:
        path = self.path()
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(self.__dict__, indent=2) + "\n", encoding="utf-8")
        # Les jetons ouvrent le compte : personne d'autre n'a à les lire.
        path.chmod(0o600)

    @classmethod
    def load(cls) -> Credentials | None:
        path = cls.path()
        if not path.is_file():
            return None
        try:
            return cls(**json.loads(path.read_text(encoding="utf-8")))
        except (json.JSONDecodeError, TypeError):
            return None

    @classmethod
    def clear(cls) -> None:
        cls.path().unlink(missing_ok=True)


def _request(url: str, payload: dict | None = None, token: str = "", method: str = "") -> Any:
    data = json.dumps(payload).encode() if payload is not None else None
    request = urllib.request.Request(
        url,
        data=data,
        method=method or ("POST" if data else "GET"),
        headers={
            "Content-Type": "application/json",
            **({"Authorization": f"Bearer {token}"} if token else {}),
        },
    )
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            body = response.read()
            return json.loads(body) if body else None
    except urllib.error.HTTPError as exc:
        # Le corps porte souvent un message utile ; s'il est illisible on se
        # rabat sur le code HTTP plutôt que de masquer l'erreur d'origine.
        detail = ""
        try:
            detail = json.loads(exc.read()).get("detail", "")
        except (json.JSONDecodeError, OSError, AttributeError):
            detail = ""
        raise SyncError(f"{exc.code} : {detail or exc.reason}") from None
    except urllib.error.URLError as exc:
        raise SyncError(f"Service injoignable : {exc.reason}") from None


def register(endpoint: str, email: str, password: str, invite_code: str = "") -> Credentials:
    body = _request(
        f"{endpoint}/v1/auth/register",
        {"email": email, "password": password, "invite_code": invite_code},
    )
    creds = Credentials(endpoint, body["access_token"], body["refresh_token"])
    creds.save()
    return account_plan(creds)


def login(endpoint: str, email: str, password: str, device_label: str = "") -> Credentials:
    body = _request(
        f"{endpoint}/v1/auth/login",
        {"email": email, "password": password, "device_label": device_label},
    )
    creds = Credentials(endpoint, body["access_token"], body["refresh_token"])
    creds.save()
    return creds


def account_plan(creds: Credentials) -> Credentials:
    """Relit l'offre du compte et la garde avec les jetons.

    Silencieux en cas d'échec : le carnet local marche hors ligne, et un
    service injoignable ne doit pas empêcher de le lire. L'offre connue reste
    alors celle de la dernière fois.
    """
    try:
        body, creds = _authorised(
            creds, lambda token: _request(f"{creds.endpoint}/v1/auth/me", token=token)
        )
    except SyncError:
        return creds

    updated = Credentials(
        creds.endpoint, creds.access_token, creds.refresh_token, body.get("plan", PLAN_FREE)
    )
    updated.save()
    return updated


def _refresh(creds: Credentials) -> Credentials:
    body = _request(f"{creds.endpoint}/v1/auth/refresh", {"refresh_token": creds.refresh_token})
    refreshed = Credentials(
        creds.endpoint, body["access_token"], body["refresh_token"], creds.plan
    )
    refreshed.save()
    return refreshed


def _authorised(creds: Credentials, call) -> tuple[Any, Credentials]:
    """Exécute un appel, en renouvelant le jeton s'il a expiré.

    Le jeton d'accès dure quinze minutes : sur un usage normal, il expire entre
    deux synchronisations. Redemander les identifiants à chaque fois serait
    intenable.
    """
    try:
        return call(creds.access_token), creds
    except SyncError as exc:
        if not str(exc).startswith("401"):
            raise
        creds = _refresh(creds)
        return call(creds.access_token), creds


def _encrypt_entry(entry: dict[str, Any], key: bytes) -> dict[str, str]:
    import os as _os

    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    from .transfer import _b64e

    nonce = _os.urandom(12)
    plain = json.dumps(entry, ensure_ascii=False, separators=(",", ":")).encode()
    cipher = AESGCM(key).encrypt(nonce, plain, None)
    return {
        "entry_id": entry["id"],
        "nonce": _b64e(nonce),
        "blob": _b64e(cipher),
        "deleted": bool(entry.get("deleted")),
    }


def _decrypt_entry(row: dict[str, str], key: bytes) -> dict[str, Any]:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    from .transfer import TransferError, _b64d

    try:
        plain = AESGCM(key).decrypt(_b64d(row["nonce"]), _b64d(row["blob"]), None)
    except Exception as exc:
        raise TransferError(
            "Déchiffrement impossible : la clef maîtresse n'est pas celle qui a "
            "servi à synchroniser ce carnet."
        ) from exc
    return json.loads(plain)


def sync(
    vault: dict[str, Any], master_key: str, creds: Credentials
) -> tuple[dict[str, Any], list[Conflict], int, Credentials]:
    """Synchronise le carnet local avec le serveur.

    Rend le carnet fusionné, les conflits, le nombre d'entrées restées sur
    l'appareil faute de place sous le plafond du compte, et la session.

    Toujours dans cet ordre : on tire d'abord, on fusionne, puis on pousse.
    Pousser sans avoir tiré écraserait ce qu'un autre appareil a écrit entre
    temps — et le serveur le refuse, précisément pour cette raison.
    """
    key = derive_transfer_key(master_key)

    pulled, creds = _authorised(
        creds, lambda token: _request(f"{creds.endpoint}/v1/vault", token=token)
    )

    remote = {"schema": 1, "updatedAt": vault.get("updatedAt", ""), "entries": []}
    for row in pulled["entries"]:
        # Un « v » résiduel est ignoré : une entrée dérive toujours en v2.
        entry = drop_version(_decrypt_entry(row, key))
        # Absent quand faux, jamais « deleted: false ». La représentation
        # canonique départage les écritures simultanées : y laisser un champ
        # que les autres implémentations n'écrivent pas ferait désigner un
        # gagnant différent selon l'appareil, et les carnets ne convergeraient
        # jamais.
        if entry.get("deleted") or row["deleted"]:
            entry["deleted"] = True
        else:
            entry.pop("deleted", None)
        remote["entries"].append(entry)

    merged, conflicts = merge(vault, remote)

    # Au-delà du plafond, le reste du carnet ne part pas : il reste propre à
    # l'appareil. Un serveur qui ne dit pas son plafond reçoit tout.
    max_entries = pulled.get("max_entries")
    if isinstance(max_entries, int) and not isinstance(max_entries, bool):
        push, local_only = select_for_push(
            merged, [row["entry_id"] for row in pulled["entries"]], max_entries
        )
    else:
        push, local_only = merged["entries"], []

    payload = {
        "base_revision": pulled["revision"],
        "entries": [_encrypt_entry(e, key) for e in push],
    }
    _, creds = _authorised(
        creds,
        lambda token: _request(f"{creds.endpoint}/v1/vault", payload, token=token),
    )

    return merged, conflicts, len(local_only), creds

File: apps/website/src/sync.ts
/**
 * Client de synchronisation.
 *
 * Le carnet est chiffré **avant** de quitter le navigateur, avec la clef de
 * transfert dérivée de la clef maîtresse. Le serveur ne reçoit que des blocs
 * opaques : il ne peut lire ni les sites, ni les identifiants.
 *
 * Les identifiants du compte sont volontairement distincts de la clef
 * maîtresse. S'authentifier avec celle-ci ferait qu'une faiblesse du service
 * exposerait les mots de passe eux-mêmes.
 */

import { deriveTransferKey } from "@/transfer";
import {
  isVaultEntryV2,
  mergeVaults,
  selectForPush,
  type Conflict,
  type Vault,
  type VaultEntry,
} from "@/vault";

export const DEFAULT_ENDPOINT = "https://thecode-api.julsql.fr";
const SESSION_KEY = "thecode.session";

export class SyncError extends Error {
  /** Statut HTTP de la réponse, 0 quand le service n'a pas répondu. */
  readonly status: number;

  constructor(message: string, status = 0) {
    super(message);
    this.status = status;
  }
}

export interface Session {
  endpoint: string;
  accessToken: string;
  refreshToken: string;
  /**
   * Offre du compte, telle que le service l'a dite la dernière fois.
   *
   * Gardée avec la session parce que la génération se fait hors ligne : sans
   * cette trace, l'écran ne saurait pas quoi proposer tant que le service
   * n'a pas répondu, et proposerait donc tout.
   */
  plan?: string;
}

function b64e(bytes: ArrayBuffer | Uint8Array): string {
  let binary = "";
  for (const b of new Uint8Array(bytes)) binary += String.fromCharCode(b);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function b64d(text: string): Uint8Array {
  const standard = text.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(standard + "=".repeat((4 - (standard.length % 4)) % 4));
  return Uint8Array.from(binary, (c) => c.charCodeAt(0));
}

export async function request(
  url: string,
  options: { payload?: unknown; token?: string; method?: string } = {},
) {
  let response: Response;
  try {
    response = await fetch(url, {
      method: options.method ?? (options.payload ? "POST" : "GET"),
      headers: {
        "Content-Type": "application/json",
        ...(options.token ? { Authorization: `Bearer ${options.token}` } : {}),
      },
      body: options.payload ? JSON.stringify(options.payload) : undefined,
    });
  } catch (e) {
    throw new SyncError(`Service injoignable : ${(e as Error).message}`);
  }

  if (!response.ok) {
    let detail = "";
    try {
      detail = ((await response.json()) as { detail?: string })?.detail ?? "";
    } catch {
      detail = "";
    }
    throw new SyncError(`${response.status} : ${detail || response.statusText}`, response.status);
  }

  return response.status === 204 ? null : response.json();
}

/**
 * La session vit dans localStorage, comme le carnet : elle ne quitte pas le
 * navigateur. Une lecture qui échoue ne doit pas casser la page.
 */
export function loadSession(): Session | null {
  try {
    const raw = localStorage.getItem(SESSION_KEY);
    return raw ? (JSON.parse(raw) as Session) : null;
  } catch {
    return null;
  }
}

export function saveSession(session: Session): void {
  try {
    localStorage.setItem(SESSION_KEY, JSON.stringify(session));
  } catch {
    // Stockage indisponible : la synchronisation de cette session marchera,
    // mais il faudra se reconnecter au prochain chargement.
  }
}

export function clearSession(): void {
  try {
    localStorage.removeItem(SESSION_KEY);
  } catch {
    // Rien à faire : la session disparaîtra avec le stockage.
  }
}

/** Ce que l'inscription réclame, avant de le demander. */
export interface RegistrationState {
  open: boolean;
  needsCode: boolean;
  /** Places restantes sans code, ou null quand la notion ne s'applique pas. */
  freeSlots: number | null;
  /**
   * Identifiant client Google, vide quand la connexion Google n'est pas
   * configurée. Donné par le service plutôt que recopié ici : deux copies
   * finiraient par ne plus correspondre, et le bouton échouerait sans raison
   * visible.
   */
  googleClientId?: string;
}

export async function registrationState(endpoint: string): Promise<RegistrationState> {
  return (await request(`${endpoint}/v1/auth/registration`)) as RegistrationState;
}

/**
 * Crée un compte.
 *
 * `code` couvre l'invitation, le parrainage et l'offre à vie : c'est le même
 * champ pour l'utilisateur, et le serveur sait lequel il tient.
 *
 * `lang` part avec l'inscription parce que le lien de vérification mène au
 * site, qui est traduit : recevoir un lien en anglais quand on lit le site en
 * français donne l'impression de s'être trompé d'endroit.
 */
export async function register(
  endpoint: string,
  email: string,
  password: string,
  code = "",
  lang = "en",
): Promise<Session> {
  const body = (await request(`${endpoint}/v1/auth/register`, {
    payload: {
      email,
      password,
      invite_code: code,
      lang,
      device_label: "site web",
    },
  })) as { access_token: string; refresh_token: string };

  return {
    endpoint,
    accessToken: body.access_token,
    refreshToken: body.refresh_token,
  };
}

export async function login(endpoint: string, email: string, password: string): Promise<Session> {
  const body = await request(`${endpoint}/v1/auth/login`, {
    payload: { email, password, device_label: "site web" },
  });
  return {
    endpoint,
    accessToken: body.access_token,
    refreshToken: body.refresh_token,
  };
}

/**
 * Ouvre la session à partir d'un jeton Google, en créant le compte au besoin.
 *
 * Une seule route pour les deux : « continuer avec Google » ne distingue pas
 * l'inscription de la connexion, et demander lequel des deux on veut
 * reviendrait à demander de se souvenir si l'on est déjà venu.
 */
export async function googleSignIn(
  endpoint: string,
  idToken: string,
  code = "",
  lang = "en",
): Promise<Session> {
  const body = await request(`${endpoint}/v1/auth/google`, {
    payload: {
      id_token: idToken,
      invite_code: code,
      lang,
      device_label: "site web",
    },
  });
  return {
    endpoint,
    accessToken: body.access_token,
    refreshToken: body.refresh_token,
  };
}

/**
 * Exécute un appel en renouvelant le jeton s'il a expiré.
 *
 * Le jeton d'accès dure quinze minutes : sur un usage normal il expire entre
 * deux synchronisations.
 */
async function withFreshToken<T>(
  session: Session,
  call: (token: string) => Promise<T>,
): Promise<{ result: T; session: Session }> {
  try {
    return { result: await call(session.accessToken), session };
  } catch (e) {
    if (!(e instanceof SyncError) || !e.message.startsWith("401")) throw e;

    const body = await request(`${session.endpoint}/v1/auth/refresh`, {
      payload: { refresh_token: session.refreshToken },
    });
    const refreshed: Session = {
      endpoint: session.endpoint,
      accessToken: body.access_token,
      refreshToken: body.refresh_token,
    };
    return { result: await call(refreshed.accessToken), session: refreshed };
  }
}

/**
 * Exécute un appel authentifié et garde la session à jour.
 *
 * Le renouvellement fait tourner le jeton : oublier de réécrire la session
 * revient à se déconnecter au prochain chargement, sans rien comprendre.
 */
export async function authorized<T>(
  session: Session,
  call: (token: string) => Promise<T>,
): Promise<T> {
  const { result, session: fresh } = await withFreshToken(session, call);
  if (fresh !== session) saveSession(fresh);
  return result;
}

async function encryptEntry(entry: VaultEntry, key: CryptoKey) {
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const plain = new TextEncoder().encode(JSON.stringify(entry));
  const cipher = await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce }, key, plain);
  return {
    entry_id: entry.id,
    nonce: b64e(nonce),
    blob: b64e(cipher),
    deleted: Boolean(entry.deleted),
  };
}

async function decryptEntry(
  row: { nonce: string; blob: string; deleted: boolean },
  key: CryptoKey,
): Promise<VaultEntry> {
  let plain: ArrayBuffer;
  try {
    plain = await crypto.subtle.decrypt(
      { name: "AES-GCM", iv: b64d(row.nonce) },
      key,
      b64d(row.blob),
    );
  } catch {
    throw new SyncError(
      "Déchiffrement impossible : la clef maîtresse n'est pas celle qui a servi à synchroniser ce carnet.",
    );
  }
  return JSON.parse(new TextDecoder().decode(plain)) as VaultEntry;
}

/**
 * Synchronise le carnet local avec le serveur.
 *
 * Toujours dans cet ordre : tirer, fusionner, pousser. Pousser sans avoir tiré
 * écraserait ce qu'un autre appareil a écrit entre temps — et le serveur le
 * refuse, précisément pour cette raison.
 */
export async function syncVault(
  vault: Vault,
  masterKey: string,
  session: Session,
): Promise<{ vault: Vault; conflicts: Conflict[]; localOnly: number; session: Session }> {
  const key = await deriveTransferKey(masterKey);

  const pulled = await withFreshToken(session, (token) =>
    request(`${session.endpoint}/v1/vault`, { token }),
  );

  const remote: Vault = { schema: 1, updatedAt: vault.updatedAt, entries: [] };
  for (const row of pulled.result.entries) {
    const entry = await decryptEntry(row, key);
    // Le carnet n'accepte que la v2 : une entrée v1 venue du serveur est
    // écartée sans erreur et ne rejoint pas le carnet local.
    if (!isVaultEntryV2(entry)) continue;
    // Absent quand faux, jamais « deleted: false ». La représentation
    // canonique départage les écritures simultanées : y laisser un champ que
    // les autres implémentations n'écrivent pas ferait désigner un gagnant
    // différent selon l'appareil, et les carnets ne convergeraient jamais.
    if (entry.deleted || row.deleted) entry.deleted = true;
    else delete entry.deleted;
    remote.entries.push(entry);
  }

  const { vault: merged, conflicts } = mergeVaults(vault, remote);

  // Au-delà du plafond, le reste du carnet ne part pas : il reste propre à
  // l'appareil. Un serveur qui ne dit pas son plafond reçoit tout.
  const { push, localOnly } =
    typeof pulled.result.max_entries === "number"
      ? selectForPush(
          merged,
          pulled.result.entries.map((row: { entry_id: string }) => row.entry_id),
          pulled.result.max_entries,
        )
      : { push: merged.entries, localOnly: [] };

  const payload = {
    base_revision: pulled.result.revision,
    entries: await Promise.all(push.map((e) => encryptEntry(e, key))),
  };
  const pushed = await withFreshToken(pulled.session, (token) =>
    request(`${session.endpoint}/v1/vault`, { payload, token }),
  );

  return { vault: merged, conflicts, localOnly: localOnly.length, session: pushed.session };
}
